### PR TITLE
Support RS256 JWTs

### PR DESCRIPTION
**Pull request by cole-h**
Fixes https://github.com/zhaofengli/attic/issues/95.

---

While this adds support for RS256 JWTs, it does not remove support for HS256 JWTs. That said, it does recommend using RS256 in documentation.

~~We switched to using the `jsonwebtoken` crate since that's what we're most familiar with, but I can probably switch back to `jwt_simple` if that is so desired.~~

### CODE REVIEW

**Comment by zhaofengli** [comment]
Hi there, thanks for the PR! Using RS256 definitely makes more sense in production scenarios. I would appreciate if `jwt_simple` is used - I switched to it from `jsonwebtoken` specifically so that the token crate works inside WebAssembly.

**Comment by cole-h** [comment]
OK, done in 0a9d4938ef482cffdc2b5e09c7d1cfaad8001f52!

File: token/src/lib.rs [hunk 1]
```diff
@@ -223,14 +224,59 @@ pub enum Error {
...
+            max_header_length: None,
+        };
+
+        match signature_type {
+            SignatureType::HS256(key) => key
+                .verify_token(token, Some(opts))
+                .map_err(Error::TokenError)
+                .map(Token),
+            SignatureType::RS256(key) => {
+                let public_key = key.public_key();
```
Comment (niklaskorz) [inline token/src/lib.rs:273]: I think it should be possible to only provide the public key to atticd, without the private key. If the public key is always derived from the private key, that kind of voids the purpose of asymmetric JWTs. Also splitting this into public key for atticd and private key for atticadm means we can user other authorities like Keycloak without exposing their private key, as the signing would be done by Keycloak itself and atticd only needs to verify.

Comment (cole-h) [reply token/src/lib.rs:273]: We actually have an implementation of this, but it's not ready to PR yet (and I don't want to block this PR on that being ready).

**Comment by fd** [comment]
This would be incredibly useful to us.

File: token/src/lib.rs (hunk 1 above)
Comment (niklaskorz) [reply token/src/lib.rs:273]: Considering that this PR is still pending two months later, maybe it's ready to include in this PR now? :)

**Comment by cole-h** [comment]
Sorry for the delay -- pubkey-only validation is now supported as of 756fef8d5f35690faf3edcf1bbe928d36ffdeb32.

**Comment by niklaskorz** [comment]
@zhaofengli Are there any more required changes for this to be merged?

**Comment by cole-h** [comment]
(Probably one thing would be addressing the conflicts, which I've done now)

**Comment by apexo** [comment]
How is this better than HS256 in the context of attic? I've left a longer comment here: https://github.com/zhaofengli/attic/issues/95#issuecomment-2371520211

**Comment by zhaofengli** [comment]
I'll be merging this in #177.

**Pull request closed without merging**